src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -0,0 +1,1341 @@
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(y)").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void TypeInference_Call()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static object F1() => new object();
+    static object? F2() => null;
+    static void G()
+    {
+        F(F1()).ToString();
+        F(F2()).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(F2()).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(F2())").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void TypeInference_Property()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static object P => new object();
+    static object? Q => null;
+    static void G()
+    {
+        F(P).ToString();
+        F(Q).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(Q).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(Q)").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void TypeInference_FieldAccess()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static object F1 = new object();
+    static object? F2 = null;
+    static void G()
+    {
+        F(F1).ToString();
+        F(F2).ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (9,9): warning CS8602: Possible dereference of a null reference.
+                //         F(F2).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F(F2)").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void TypeInference_Literal()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G()
+    {
+        F(0).ToString();
+        F('A').ToString();
+        F(""B"").ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void TypeInference_Default()
+        {
+            var source =
+@"class C
+{
+    static T F<T>(T t) => t;
+    static void G()
+    {
+        F(default(object)).ToString();
+        F(default(int)).ToString();
+        F(default(string)).ToString();
```

Comment:
Maybe worthwhile to add a case for `F(default).ToString();`. The inference should fail, but we don't expect any nullability-related diagnostics.

Reply:
Added.